docs/design/design-patterns/rest-api-design-guidance/creating-api-contract.md

```diff
@@ -0,0 +1,56 @@
+- Server side boilerplate code e.g. controllers, DTOs etc. can be auto generated from API contracts.
+- May improve collaboration between API producer & consumer teams.
+
+Planning a Design-First Development:
+
+1. Identify use cases & key services which API should offer.
+2. Identify key stakeholders of API and try to include them during API design phase to get continuous feedback.
+3. Write API contract definitions.
+4. Maintain consistent style for API status codes, versioning, error responses etc.
+5. Encourage peer reviews via pull requests.
+6. Generate server side boilerplate code & client SDKs from API contract definitions.
+
+Important Points to consider:
+
+- If API requirements changes often during initial development phase, than a Design-First approach may not be a good fit as this will introduce an overhead to always first update & maintain API contract definitions again & again.
+- It might be worthwhile to first try out your platform specific code generator and evaluate how much more additional work will be required in order to meet your project requirements and coding guidelines because it is possible that a particular platform specific code generator might not be able to generate a flexible & maintainable implementation of actual code. For instance If your web framework requires annotations to be present on your controller classes (e.g. for API versioning or authentication), make sure that the code generation tool you use fully supports them.
+
+## Code-First Approach
+
+A Code-First approach means that development teams first implements server side api interface code e.g. controllers, DTOs etc. and than generates API contract definitions out of it. In current times this approach is more widely popular within developer community than Design-First Approach.
+
+This approach has advantage if team wants to quickly implement APIs and also provides flexibility to react very quickly on an unexpected API requirement change.
+
+Key Benefits of this approach:
+
+- Rapid development of APIs as development team can start implementing APIs much faster directly after understanding key requirements & use cases.
+- Development team has better control & flexibility to implement server side api interfaces in a way which best suited for project structure.
+- More popular among development teams so its easier to get consensus on a related topic and also has more ready to use code examples available on various blogs or developer forums regarding how to generate Open API definitions out of actual code.
+- During initial phase of development where both API producer & consumers requirements might change often this approach is better as it provides flexibility to quickly react on such changes.
+
+Important Points to consider:
+
+- A generated Open API definition can be outdated, so its important to have automated checks to avoid this otherwise generated client SDKs will be out of sync and may cause issues for API consumers.
```

Comment:
```suggestion
- A generated Open API definition can become outdated, so its important to have automated checks to avoid this otherwise generated client SDKs will be out of sync and may cause issues for API consumers.
```